### DISCUSSION TITLE
merge function issue

### DISCUSSION THREAD

**Discussion by marionTH** [mtex-toolbox/mtex#1372] [Ask Anything]
Hello,

I have a problem with the merge function. 

The first part of my matlab code is quite perfectly working .
```
[grains,ebsd.grainId,ebsd.mis2mean] = calcGrains(ebsd,'angle',HAGB_threshold);
[grains_Indexed,ebsd_Indexed.grainId,ebsd_Indexed.mis2mean] = calcGrains(ebsd('indexed'),'angle',HAGB_threshold);
ebsd(grains(grains.grainSize<thresh_set))= [];
[grains_Indexed_SizeFiltered,ebsd_Indexed_SizeFiltered.grainId,ebsd_Indexed_SizeFiltered.mis2mean] = calcGrains(ebsd('indexed'),'angle',HAGB_threshold);
grains_Indexed_SizeFiltered = smooth(grains_Indexed_SizeFiltered,5);
```
I decided to remove the twining from the grains_Indexed_SizeFiltered in the following way : 
```
gB=grains_Indexed_SizeFiltered.boundary;
gB_mainPhase = gB(mainPhase_id,mainPhase_id);
%% Sigma3 twin boundary
twinning = rotation('axis',vector3d(1,1,1),'angle',twinAngle*degree);
isTwinning = angle(gB_mainPhase.misorientation,twinning)<twinTolAngle*degree;
twinBoundary = gB_mainPhase(isTwinning);
%% mergedGrains = grains wo twin boundary

[mergedGrains,parentId] = merge(grains_Indexed_SizeFiltered,twinBoundary);
```
This first use of the merge function works fine. 

Then i want to selected my recrystallised grains by a double critera that is : 
```
Condition_REX = zeros(length(grains_Indexed_SizeFiltered), 1);
for i=1:length(Condition_REX)
    if ((grains_Indexed_SizeFiltered(i).GOS*(180/pi) <= GOS_REX_threshold) & (GAKAM1(i) <= GAKAM_REX_threshold))
        Condition_REX(i)=1.0;
    else
        Condition_NONREX(i)=1.0;
    end
end

grains_REX = grains_Indexed_SizeFiltered(Condition_REX == 1.0);
```

But once i want to use the merge function on grains_REX  like this this is where the trouble begin.
```
twinBoundary_REX = gB_mainPhase_REX(isTwinning);
```

I am aware of the bug that as been fixed on the new mtex version concerning the negative values , that was my first problem. But now i have a different issue and i get this error line : 

Error using sparse
Index exceeds array bounds.

Error in grain2d/merge (line 68)
    A = A | sparse(mergeId(:,1),mergeId(:,2),1,maxId+1,maxId+1);

Error in post_traitment_EBSD_torsion_marion (line 431)
[merged,parents] = merge(grains_REX,twinBoundary_REX);

Did anyone had any similar probleme ?

**Comment by kilir** [comment]
Hi,
can it be that the set of grains `grains_REX` and your `twinBoundary_REX` do not belong together? From the code you provide, it is not clear where `gB_mainPhase_REX` comes. 
Cheers,
Rüdiger

**Comment by marionTH** [comment]
I think i circle more precisly my issue. I selected with the given criteria my recrystallized grains. And what i want to do next is remove the twin boundaries inside them. I think that many of my trouble is coming from the fact that some 60° degrees bound 


De: "Rüdiger Kilian" ***@***.***> 
À: "mtex-toolbox/mtex" ***@***.***> 
Cc: "Marion ROTH" ***@***.***>, "Author" ***@***.***> 
Envoyé: Dimanche 13 Mars 2022 16:21:34 
Objet: Re: [mtex-toolbox/mtex] merge function issue (Discussion #1372) 





Hi, 
can it be that the set of grains grains_REX and your twinBoundary_REX do not belong together? From the code you provide, it is not clear where gB_mainPhase_REX comes. 
Cheers, 
Rüdiger 

— 
Reply to this email directly, [ https://github.com/mtex-toolbox/mtex/discussions/1372#discussioncomment-2350383 | view it on GitHub ] , or [ https://github.com/notifications/unsubscribe-auth/AYFYNBHB66DYIG5NZZTHC4DU7YBX5ANCNFSM5QQD4P4A | unsubscribe ] . 
Triage notifications on the go with GitHub Mobile for [ https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675 | iOS ] or [ https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub | Android ] .

**Reply by marionTH** [reply]
Here some more content to understand my problem,
![merge issue ](https://user-images.githubusercontent.com/101418628/158197558-2a7f93d6-8917-4a35-9466-31c8b89da797.jpg)
 Above, is a image of my microstrucutre, in blue are the REX grains that i have selected with the criterion that i mentoned earlier. In red these are the  all  the boundaries related to these REX grains. In green it is all the twin boundaries related to REX grains and finally in blue we have the twin boundaries inside REX grains. Here is how a calculated these later boundaries  : 
```
gB_REX=grains_REX.boundary;
gB_mainPhase_REX = gB_REX(mainPhase_id,mainPhase_id);
twinning = rotation('axis',vector3d(1,1,1),'angle',twinAngle*degree);
ag=angle(gB_mainPhase_REX.misorientation,twinning);
isTwinning_REX =angle(gB_mainPhase_REX.misorientation,twinning)<=twinTolAngle*degree;
twinBoundary_REX = gB_mainPhase_REX(isTwinning_REX);
isrecbound = ismember(gB_mainPhase_REX.grainId, grains_REX.id);
%both columns must be one
bothRec=isrecbound(:,1) & isrecbound(:,2);
bothRec = [bothRec, bothRec];


figure()
plot(ebsd,ebsd.bc);
mtexColorbar('title','REX grains (°)','location','eastoutside')
colormap gray %in grey scale
hold on 
plot(grains_REX)
hold off 
hold on 
plot(gB_REX, 'linecolor', 'r','linewidth',2)
hold off 
hold on 
plot(twinBoundary_REX, 'linecolor', 'g','linewidth',2)
hold off 
hold on 
plot(gB_mainPhase_REX(isTwinning_REX & bothRec ), 'linecolor', 'b','linewidth',2)
hold off 

[mergedGrainsrec, parentId] = merge(grains_REX, gB_mainPhase_REX(isTwinning_REX & bothRec ));
```
Even with the blue selection; i have some boundaries that are not complete such as :
![uncomplete boundaries](https://user-images.githubusercontent.com/101418628/158198834-34d1d6f3-6291-4cc9-8365-faef76e735c6.png)

And i think this is was is explaining the erreor message that i get : 

Index exceeds the number of array elements (1807).

Error in grain2d/merge (line 195)
grainsMerged.boundary.grainId(ind) = old2newId(grains.boundary.grainId(ind));

Does anyone has an idea of how i can select entierly theses boundaries in order to be able to merge my REX grains correctly ?
